k2/python/k2/fsa.py

```diff
@@ -599,43 +685,107 @@ def from_dict(cls, dict_in: Dict[str, Any]) -> 'Fsa':
-          This is an in-place operation.
-
         Args:
           device:
             An instance of `torch.device`. It supports only cpu and cuda.
 
         Returns:
-          Return `self`.
+          Returns a new Fsa which is this object copied to the given device
+         (or this object itself, if the device was the same)
         '''
+        # Keep this code in sync with that in clone()
         assert device.type in ('cpu', 'cuda')
-        if device.type == 'cpu' and self.is_cpu():
-            return self
-        elif device.type == 'cuda' and self.is_cuda():
+        if device == self.scores.device:
             return self
 
-        self.arcs = self.arcs.to(device)
+        ans = Fsa(self.arcs.to(device), properties=self.properties)
 
-        for name, value in self.named_tensor_attr():
-            setattr(self, name, value.to(device))
+        for name, value in self.named_tensor_attr(include_scores = False):
+            setattr(ans, name, value.to(device))
 
-        self._grad_cache = OrderedDict()
+        for name, value in self.named_non_tensor_attr():
+            setattr(ans, name, value)
 
-        return self
+        # Don't copy members of self._cache.  They don't all have convenient
+        # .to() methods.
+
+        # The following is a magic invocation to make sure
+        # the backprop happens.
+        phantom_set_scores_to(ans, self.scores.to(device))
+
+        return ans
+
+    def clone(self) -> 'Fsa':
+        """
+        Return an Fsa that is a clone of this one, i.e. a close approximation
+        to what you'd get if you did .clone() on all its tensor members.
+        Any non-tensor attributes are copied over.
+        """
+        # Keep this code in sync with that in to()
+        ans = Fsa(self.arcs.clone(), properties=self.properties)
+
+        for name, value in self.named_tensor_attr(include_scores = False):
+            setattr(ans, name, value.clone())
+
+        for name, value in self.named_non_tensor_attr():
+            setattr(ans, name, value)
+
+        # Just copy elements of the _cache that we might already have..
+        # These don't directly participate in autograd, and are not supposed to
+        # be modified by the user, so this should be safe (i.e. it should
+        # be safe to do this without clone(); these are mostly not tensors
+        # anyway.
+        for name, value in self._cache:
+            ans._cache[name] = value
+
+        # The following is a magic invocation to make sure
+        # the backprop happens.
+        phantom_set_scores_to(ans, self.scores)
 
-    def named_tensor_attr(self) -> Iterator[Tuple[str, torch.Tensor]]:
+        return ans
+
+    def detach(self) -> 'Fsa':
+        """
+        Return an Fsa that shares the underlying data with this one,
+        except gradients are not tracked.
+        Any non-tensor attributes are copied over.
+        """
+        ans = Fsa(self.arcs.detach(), properties=self.properties)
```

Comment:
So where is the method `detach` of `arcs` defined?

Reply:
Ah yes, detach() is not needed on arcs, it's not a tensor.